scripts/get-workspace-map.sh

```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
+# get-workspace-map.sh [--stage] [--prod] [context-1 context-2... context-n]
+#   Read Namespace resources and generate a JSON object mapping from namespaces
+#   to workspace based on the value of label toolchain.dev.openshift.com/space.
+#
+#   This script assumes `oc` is preconfigured with all the required clusters
+#   and could be found in $PATH.
+#
+#   if --stage or --prod provided, the script will look for contexts matching
+#   staging and production clusters.
+#   Any other argument will be considered as an explicit context name.
+#
+set -o pipefail -o errexit -o nounset
+
+# Parse command line arguments
+EXPLICIT=""
+for i in "$@"
+do
+case $i in
+    --stage)
+    STAGE=true
+    ;;
+    --prod)
+    PROD=true
+    ;;
+    *)
+    EXPLICIT="${i} ${EXPLICIT}"
+    ;;
+esac
+done
+
+STAGE=${STAGE:-""}
+PROD=${PROD:-""}
+
+if [[ -n $STAGE ]]; then
+# Take all non-host staging clusters and keep only their names
+  STAGE=$(
+    oc config get-contexts \
+      | { grep -E 'api-stone-stg\S+' || test $? = 1; } \
+      | { grep -v api-stone-stg-host || test $? = 1; } \
+      | sed 's/^[ \*]*//g' \
+      | sed 's/ .*$//g' \
+      | tr '\n' ' '
+  )
+  if [[ -z $STAGE ]]; then
+    echo "No staging clusters found."
+  fi
+fi
+
+if [[ -n $PROD ]]; then
+# Take all non-host production clusters and keep only their names
+  PROD=$(
+    oc config get-contexts \
```

Comment:
all the `grep`s and `sed`s can probably be combined to a single `sed` run. For example here is how you implement a simple grep with sed:
```
sed -n '/regex-here/p'
```
Where `p` is a simple print command but there are other commands such as `d` that can be used to filter out lines, and the well-known `s`.

I don't think the `tr` is needed either because the following would make base get rid on the newlines for us:
```
FOO=($(foo | bar))

Reply:
I'll give it a try. I don't like it when there's too much magic involved. Hard to troubleshoot.

Reply:
I've been looking into replacing those multiple `grep` and `sed` calls to a single `sed` or even at replacing parts of it. I was not able to get it to work yet, but I think the result would be less maintainable than the current approach.

regex and sed already have the potential of becoming write-only code. I'd prefer leaving it as a sequence of simple steps rather than a single harder-to-manage step.

Reply:
I do want to find a way to simplify this - I think this logic for getting the clusters is occupying too much code ATM - perhaps we can leave only QUERY_URLS and drop the rest while making it default to the stage hosts?

Also I think we need a way to hard-wire the `contexts` variable for the automated use case where we control the contents of the kubeconfig file.

Reply:
> I do want to find a way to simplify this - I think this logic for getting the clusters is occupying too much code ATM - perhaps we can leave only QUERY_URLS and drop the rest while making it default to the stage hosts?

Not sure what you mean.

> Also I think we need a way to hard-wire the contexts variable for the automated use case where we control the contents of the kubeconfig file.

you mean that by default the script will use all contexts in the kubeconfig file?

Reply:
Turns out there's a `kubectl config view` that can be processed with `-o=go-template`

Reply:
nice! 

lets setup a few minutes to meet and iron out the behaviour around this

Reply:
@ifireball please take another look.